app/src/main/java/io/homeassistant/companion/android/widgets/graph/GraphWidget.kt

```diff
@@ -0,0 +1,388 @@
+                setAvoidFirstLastClipping(true)
+                isAutoScaleMinMaxEnabled = true
+            }
+
+            axisLeft.apply {
+                setDrawGridLines(true)
+                textColor = dynTextColor
+                textSize = 12F
+            }
+
+            axisRight.apply {
+                setDrawGridLines(false)
+                setDrawLabels(false)
+            }
+
+            legend.apply {
+                isEnabled = true
+                textColor = dynTextColor
+                textSize = 12F
+                verticalAlignment = Legend.LegendVerticalAlignment.BOTTOM
+                horizontalAlignment = Legend.LegendHorizontalAlignment.RIGHT
+                orientation = Legend.LegendOrientation.HORIZONTAL
+                setDrawInside(false)
+            }
+
+            description = Description().apply {
+                text = "$timeRange h"
+                textColor = dynTextColor
+            }
+
+            legend.isEnabled = true
+            description.isEnabled = true
+        }
+
+        val mainGraphColor = ContextCompat.getColor(context, commonR.color.colorPrimary)
+
+        val dataSet = LineDataSet(entries, label).apply {
+            color = mainGraphColor
+            lineWidth = 2F
+            circleRadius = 1F
+            setDrawCircleHole(false)
+            setCircleColor(mainGraphColor)
+            mode = LineDataSet.Mode.CUBIC_BEZIER
+            setDrawCircles(true)
+            setDrawValues(false)
+        }
+
+        lineChart.data = LineData(dataSet)
+
+        lineChart.layout(0, 0, width, height)
+
+        return lineChart
+    }
+
+    override suspend fun getAllWidgetIdsWithEntities(context: Context): Map<Int, Pair<Int, List<String>>> =
+        repository.getAllFlow()
+            .first()
+            .associate { it.id to (it.serverId to listOf(it.entityId)) }
+
+    private suspend fun resolveTextToShow(
+        serverId: Int,
+        entityId: String?,
+        suggestedEntity: Entity<Map<String, Any>>?
+    ): ResolvedText {
+        var entity: Entity<Map<String, Any>>? = null
+        var entityCaughtException = false
+        try {
+            entity = if (suggestedEntity != null && suggestedEntity.entityId == entityId) {
+                suggestedEntity
+            } else {
+                entityId?.let { serverManager.integrationRepository(serverId).getEntity(it) }
+            }
+        } catch (e: Exception) {
+            Log.e(TAG, "Unable to fetch entity", e)
+            entityCaughtException = true
+        }
+
+        return ResolvedText(entity?.friendlyName, entityCaughtException)
+    }
+
+    private suspend fun fetchHistory(appWidgetId: Int, serverId: Int, entityId: String, fromMillis: Long, toMillis: Long) {
+        val entitiesStatesList: List<List<Entity<Map<String, Any>>>>?
+        try {
+            entitiesStatesList = serverManager.integrationRepository(serverId)
+                .getHistory(
```

Comment:
Unless I'm missing something, this is the only place you're getting history from the API. This function (`fetchHistory`) is only called when the widget is added. That means that the issue we previously pointed out (the device being responsible for getting the history - device not connected means no history updates) still exists.

Shouldn't this be called any time after the widget has lost connection?

Reply:
Hello! @jpelgrom I have been testing this behaviour and, I could not reproduce your scenario by having the widget almost 24h. When I unlock and check the widget it updates, maybe not as often but the idea is also to not hammer by downloading everytime soo much data.

Reply:
Do you think is a good idea to add a 'cron' ?

Reply:
That doesn't seem necessary for a widget. The history API allows you to fetch any missing data, why not use that to get data from the last update (or start of time range, if later) until now on screen on? If there is nothing, the data usage is minimal.